### PR TITLE
[Merged by Bors] - feat: port LinearAlgebra.Contraction 

### PR DESCRIPTION
**Pull request by mattrobball**
---

- [x] depends on: #3659
[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by semorrison** [comment]
This PR/issue depends on:
* leanprover-community/mathlib4#3659
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by mattrobball** [comment]
Things left to do:
- [ ] align names in the comments with #naming4
- [ ] decide the naming convention for left tensor, Ltensor or LTensor, and right tensor
- [ ] diagnose the timeouts in the failing simps

**Comment by mattrobball** [comment]
I won't be able to touch it for a bit probably.

File: Mathlib/LinearAlgebra/Contraction.lean [hunk 1]
```diff
@@ -0,0 +1,347 @@
...
+-/
+
+-- Porting note: universe metavariables behave oddly
+universe w u v₁ v₂ v₃ v₄
+
+variable {ι : Type w} (R : Type u) (M : Type v₁) (N : Type v₂)
+  (P : Type v₃) (Q : Type v₄)
+
+-- Porting note: this does not appear to work; see below
+attribute [local ext] TensorProduct.ext
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/Contraction.lean:33]: `local ext high` works here.

Comment (mattrobball) [reply Mathlib/LinearAlgebra/Contraction.lean]: Thanks for figuring that out. I updated the porting note here and removed the manual ext lemma calls.

**Comment by semorrison** [comment]
This PR/issue depends on:
* ~~leanprover-community/mathlib4#3659~~
By **[Dependent Issues](https://github.com/z0al/dependent-issues)** (🤖). Happy coding!

**Comment by semorrison** [comment]
I merged `master` to get a clean diff.

File: Mathlib/LinearAlgebra/Contraction.lean [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+-/
+
+-- Porting note: universe metavariables behave oddly
+universe w u v₁ v₂ v₃ v₄
+
+variable {ι : Type w} (R : Type u) (M : Type v₁) (N : Type v₂)
+  (P : Type v₃) (Q : Type v₄)
+
+-- Porting note: apparently `high` is necessary for this to fire below, odd
+attribute [local ext high] TensorProduct.ext
```
Comment (semorrison) [inline Mathlib/LinearAlgebra/Contraction.lean:33]: The problem is not that it doesn't fire, but just that a different `@[ext]` lemma fires first.

**Comment by semorrison** [comment]
Otherwise looks good!

bors d+

File: Mathlib/LinearAlgebra/Contraction.lean (hunk 2 above)
Comment (mattrobball) [reply Mathlib/LinearAlgebra/Contraction.lean:33]: D'oh!

**Comment by mattrobball** [comment]
bors merge

**Pull request closed without merging**